Fix #10506 by adding `ExpectedNonNull` `ShellError` if required field is entered as `$nothing`, `null`, "", etc.

brunerm99 opened the pull request:
<!--
if this PR closes one or more issues, you can automatically link the PR with
them by using one of the [*linking keywords*](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword), e.g.
- this PR should close #xxxx
- fixes #xxxx

you can also mention related issues, PRs or discussions!
-->

# Description
<!--
Thank you for improving Nushell. Please, check our [contributing guide](../CONTRIBUTING.md) and talk to the core team before making major changes.

Description of your pull request goes here. **Provide examples and/or screenshots** if your changes affect the user experience.
-->
Fix #10506 by adding `ExpectedNonNull` `ShellError` if required field is entered as `$nothing`, `null`, " ", etc.

This adds a new `ShellError`, `ExpectedNonNull`, taking the expected type and span.

# User-Facing Changes
<!-- List of all changes that impact the user experience here. This helps us keep track of breaking changes. -->
Will get a more helpful error in the case described by #10506. Examples:
```nushell
➜ {scheme: "", host: "github.com"} | url join
Error: nu::shell::expected_non_null

  × Expected string found null.
   ╭─[entry #16:1:1]
 1 │ {scheme: "", host: "github.com"} | url join
   ·          ─┬
   ·           ╰── expected string, found null
   ╰────
```

```nushell
❯ {scheme: "https", host: null} | url join
Error: nu::shell::expected_non_null

  × Expected string found null.
   ╭─[entry #19:1:1]
 1 │ {scheme: "https", host: null} | url join
   ·                         ──┬─
   ·                           ╰── expected string, found null
   ╰────
```

# Tests + Formatting
<!--
Don't forget to add tests that cover your changes.

Make sure you've run and fixed any issues with these commands:

- `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used` to check that you're using the standard code style
- `cargo test --workspace` to check that all tests pass (on Windows make sure to [enable developer mode](https://learn.microsoft.com/en-us/windows/apps/get-started/developer-mode-features-and-debugging))
- `cargo run -- -c "use std testing; testing run-tests --path crates/nu-std"` to run the tests for the standard library

> **Note**
> from `nushell` you can also use the `toolkit` as follows
> ```bash
> use toolkit.nu  # or use an `env_change` hook to activate it automatically
> toolkit check pr
> ```
-->
All pass.

brunerm99 commented:
@sholderbach from #10506 
> Has this been fully resolved by https://github.com/nushell/nushell/pull/10565?

No, but this should.

sholderbach commented:
Will look into the details tomorrow but in general we don't treat the empty string as null.

brunerm99 commented:
Yeah probably shouldn't group it in with null. I could change it to say something like: `expected string, found ""`. 

That will have to be tomorrow though.

amtoine commented:
i agree with @sholderbach that the first error would be less confusing if it were to say
```nushell
➜ {scheme: "", host: "github.com"} | url join
Error: nu::shell::unexpected_argument

  × Expected non-empty string.
   ╭─[entry #16:1:1]
 1 │ {scheme: "", host: "github.com"} | url join
   ·          ─┬
   ·           ╰── expected non-empty string
   ╰────
```
maybe creating another error type just for this is not the right way to go :thinking: 

the second error looks very nice :ok_hand:

brunerm99 commented:
Updated examples:
```nushell
➜ {scheme: "https", host: ""} | url join
Error: nu::shell::unsupported_config_value

  × Unsupported config value
   ╭─[entry #1:1:1]
 1 │ {scheme: "https", host: ""} | url join
   ·                         ─┬
   ·                          ╰── expected non-empty string, got empty string
   ╰────
```
```nushell
❯ {scheme: null, host: "github.com"} | url join
Error: nu::shell::unsupported_config_value

  × Unsupported config value
   ╭─[entry #2:1:1]
 1 │ {scheme: null, host: "github.com"} | url join
   ·          ──┬─
   ·            ╰── expected non-empty string, got null or empty value
   ╰────
```

```nushell
➜ {scheme: [], host: "github.com"} | url join
Error: nu::shell::unsupported_config_value

  × Unsupported config value
   ╭─[entry #1:1:1]
 1 │ {scheme: [], host: "github.com"} | url join
   ·          ─┬
   ·           ╰── expected non-empty string, got null or empty value
   ╰────
```

I went for the `got null or empty value` on the second example because in the code currently the only two paths to a failure on matching the value is 
1. an empty string (example 1)
2. anything else which is "empty" (examples 2,3)

and I didn't want to add much more logic for a simple error.

I also removed the new `ShellError` type.

amtoine commented:
mm now that we can see it, i'm personally not a huge fan of these... :thinking: 

if `scheme` and `host` and the rest can only be non-empty `string`s, then maybe `url join` should check the real `Value` and then if it's a `Value::String`, whether or not it's the empty one :innocent:

brunerm99 commented:
@amtoine At this point in the `add_component` function the `Value` must be able to be converted to a string via `Value::as_string()` or else it continues onto the next input. 

Are you saying check all values for being non-empty strings and error on any if they are empty (instead of just on `host` and `scheme`)? Or check within just within the `host` and `scheme` cases for emptiness of the string?

amtoine commented:
this might be a bit too far out of scope for this PR and the work you've done :thinking: 

but what i mean is that, to be as little confusing as possible for the user, i think we should have errors like
- when there is a real missing config value (already like this right now)
```nushell
❯ {host: "github.com"} | url join
Error: nu::shell::missing_parameter

  × Missing parameter: scheme.
   ╭─[entry #1:1:1]
 1 │ {host: "github.com"} | url join
   · ──────────┬─────────
   ·           ╰── missing parameter: scheme
   ╰────
```
- when the type is incorrect, i.e. non-string (similar to what we have right now, even though more precise)
```nushell
❯ {scheme: null, host: "github.com"} | url join
Error: nu::shell::unsupported_config_value

  × Unsupported config value
   ╭─[entry #2:1:1]
 1 │ {scheme: null, host: "github.com"} | url join
   ·          ──┬─
   ·            ╰── expected string, got nothing
   ╰────
```
```nushell
➜ {scheme: [], host: "github.com"} | url join
Error: nu::shell::unsupported_config_value

  × Unsupported config value
   ╭─[entry #1:1:1]
 1 │ {scheme: [], host: "github.com"} | url join
   ·          ─┬
   ·           ╰── expected string, got list
   ╰────
```
- when the type is correct but the string is empty
```nushell
➜ {scheme: "https", host: ""} | url join
Error: nu::shell::unsupported_config_value

  × Unsupported config value
   ╭─[entry #1:1:1]
 1 │ {scheme: "https", host: ""} | url join
   ·                         ─┬
   ·                          ╰── expected non-empty string
   ╰────
```

brunerm99 commented:
@amtoine I looked into it and this would require some matching of the value after checking if it can be converted to a string using `Value.as_string()`. `.as_string()` already does this matching so I figured I would try just using it's error (see below). I think the message is clear about what it expects and it requires no more code to create error messages. I can also do it with the message you said but this seems simpler and doesn't add anything unnecessary. Thoughts? 

```nushell
❯ {scheme: hello, host: null} | url join
Error: nu::shell::cant_convert

  × Can't convert to string.
   ╭─[entry #3:1:1]
 1 │ {scheme: hello, host: null} | url join
   ·                       ──┬─
   ·                         ╰── can't convert nothing to string
   ╰────
```

```nushell
❯ {scheme: hello, host: []} | url join
Error: nu::shell::cant_convert

  × Can't convert to string.
   ╭─[entry #5:1:1]
 1 │ {scheme: hello, host: []} | url join
   ·                       ─┬
   ·                        ╰── can't convert list<any> to string
   ╰────
```

amtoine commented:
i think this looks quite good, because you'd see what the actual and expected types are :ok_hand: 

let's go with this, if you have time to make the change, that'd be amazing :pray: 

if the wording is not perfect in the end, this can be addressed in a follow-up PR :wink:

brunerm99 commented:
Ok... I hope this commit finalizes this PR.

The updated error messages are as follows:

1. Input incorrect type:
```nushell
➜ {scheme: hello, host: []} | url join
Error: nu::shell::cant_convert

  × Can't convert to string.
   ╭─[entry #2:1:1]
 1 │ {scheme: hello, host: []} | url join
   ·                       ─┬
   ·                        ╰── can't convert list<any> to string
   ╰────
```

2. Empty String passed for required input (`host`, `scheme`)
```nushell
❯ {scheme: hello, host: ""} | url join
Error: nu::shell::unsupported_config_value

  × Unsupported config value
   ╭─[entry #3:1:1]
 1 │ {scheme: hello, host: ""} | url join
   ·                       ─┬
   ·                        ╰── expected non-empty string, got empty string
   ╰────
```

3. Missing required input
```nushell
❯ {scheme: hello} | url join
Error: nu::shell::missing_parameter

  × Missing parameter: host.
   ╭─[entry #4:1:1]
 1 │ {scheme: hello} | url join
   · ───────┬───────
   ·        ╰── missing parameter: host
   ╰────
```

amtoine commented:
@brunerm99 
yeah i think this looks much better than what we have now :pray: 
if you can fix the CI, then it looks like it's very close to land :wink:

brunerm99 commented:
Hey @amtoine, I fixed the problem but it looks like the CI failed due to a network error. Snip from failed test: 
```
Error: nu::shell::network_failure

  × Network failure
   ╭─[source:1:1]
 1 │ http get --insecure https://expired.badssl.com/
   ·                     ─────────────┬─────────────
   ·                                  ╰── Cannot make request to https://expired.badssl.com/, there was an error establishing a connection.
   ╰────
```

Could this CI be re-run? Looks unrelated to my code, I think.

brunerm99 commented:
@amtoine CI is now passing. Thanks @sholderbach for restarting it.

amtoine commented:
oh yeah i got the exact same failure in https://github.com/nushell/nushell/pull/10341 :open_mouth:

amtoine commented:
let's land this PR, it looks much better than before :pray:

brunerm99 commented:
@amtoine 

> oh yeah i got the exact same failure in #10341 😮

I was so confused for a minute :laughing: 

Thanks for the help!

amtoine commented:
> I was so confused for a minute 😆
> 
same because it was so unrelated to the changes :laughing: 

> Thanks for the help!
>
anytime :blush:
